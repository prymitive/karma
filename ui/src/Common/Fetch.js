import merge from "lodash.merge";

import promiseRetry from "promise-retry";

const CommonOptions = {
  mode: "cors",
  credentials: "include",
  redirect: "follow",
};

const FetchRetryConfig = {
  retries: 9,
  minTimeout: 2000,
  maxTimeout: 5000,
};

const FetchGet = async (uri, options, beforeRetry) =>
  await promiseRetry(
    (retry, number) =>
      fetch(
        uri,
        merge(
          {},
          {
            method: "GET",
          },
          CommonOptions,
          {
            mode: number <= FetchRetryConfig.retries ? "cors" : "no-cors",
          },
          options
        )
      ).catch((err) => {
        beforeRetry && beforeRetry(number);
        return retry(err);
      }),
    FetchRetryConfig
  );

export { CommonOptions, FetchGet, FetchRetryConfig };
