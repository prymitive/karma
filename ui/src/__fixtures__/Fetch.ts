import { MockAlert, MockAlertGroup, MockSilence } from "__fixtures__/Alerts";
import {
  APIAlertsResponseT,
  APILabelCounterT,
  APIManagedSilenceT,
} from "Models/APITypes";

const EmptyAPIResponse = (): APIAlertsResponseT => ({
  status: "success",
  error: "",
  timestamp: new Date().toISOString(),
  version: "fakeVersion",
  upstreams: {
    counters: { total: 1, healthy: 1, failed: 0 },
    clusters: { default: ["default"] },
    instances: [
      {
        name: "default",
        cluster: "default",
        clusterMembers: ["default"],
        uri: "http://localhost",
        publicURI: "http://localhost",
        error: "",
        version: "0.21.0",
        readonly: false,
        corsCredentials: "include",
        headers: {},
      },
    ],
  },
  silences: { default: {} },
  grids: [],
  receivers: ["by-cluster-service", "by-name"],
  totalAlerts: 0,
  colors: {},
  filters: [
    {
      text: "label=value",
      name: "label",
      matcher: "=",
      value: "value",
      hits: 0,
      isValid: true,
    },
  ],
  counters: [] as APILabelCounterT[],
  settings: {
    sorting: {
      grid: {
        order: "startsAt",
        reverse: false,
        label: "alertname",
      },
      valueMapping: {
        cluster: {
          dev: 3,
          prod: 1,
          staging: 2,
        },
        severity: {
          critical: 1,
          info: 3,
          warning: 2,
        },
      },
    },
    silenceForm: {
      strip: {
        labels: [],
      },
    },
    alertAcknowledgement: {
      enabled: false,
      durationSeconds: 900,
      author: "karma / author missing",
      comment: "ACK! Mock comment",
    },
    staticColorLabels: ["job"],
    annotationsDefaultHidden: false,
    annotationsHidden: [],
    annotationsVisible: [],
    annotationsEnableHTML: false,
    historyEnabled: true,
  },
  authentication: {
    username: "",
    enabled: false,
  },
});

const MockAPIResponse = (): APIAlertsResponseT => {
  const response = EmptyAPIResponse();
  response.grids = [
    {
      labelName: "",
      labelValue: "",
      alertGroups: [
        MockAlertGroup(
          { alertname: "foo" },
          [MockAlert([], { instance: "foo" }, "suppressed")],
          [],
          { cluster: "dev" },
          {}
        ),
      ],
      stateCount: {
        unprocessed: 1,
        suppressed: 2,
        active: 3,
      },
    },
  ];
  response.totalAlerts = 25;
  return response;
};

const MockSilenceResponse = (
  cluster: string,
  count: number
): APIManagedSilenceT[] => {
  const silences: APIManagedSilenceT[] = [];
  for (let index = 1; index <= count; index++) {
    const silence = MockSilence();
    silence.id = `silence${index}`;
    silences.push({
      cluster: cluster,
      alertCount: 123,
      isExpired: index < 3,
      silence: silence,
    });
  }
  return silences;
};

export { EmptyAPIResponse, MockAPIResponse, MockSilenceResponse };
