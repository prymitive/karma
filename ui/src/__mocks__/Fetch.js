import { MockAlert, MockAlertGroup, MockSilence } from "./Alerts";

const EmptyAPIResponse = () => ({
  status: "success",
  timestamp: new Date().toISOString(),
  version: "fakeVersion",
  upstreams: {
    counters: { total: 1, healthy: 1, failed: 0 },
    instances: [{ name: "default", uri: "file:///mock", error: "" }],
  },
  silences: { default: {} },
  grids: [],
  receivers: ["by-cluster-service", "by-name"],
  totalAlerts: 0,
  colors: {},
  filters: [
    {
      text: "label=value",
      name: "label",
      matcher: "=",
      value: "value",
      hits: 0,
      isValid: true,
    },
  ],
  counters: {},
  settings: {
    sorting: {
      grid: {
        order: "startsAt",
        reverse: false,
        label: "alertname",
      },
      valueMapping: {
        cluster: {
          dev: 3,
          prod: 1,
          staging: 2,
        },
        severity: {
          critical: 1,
          info: 3,
          warning: 2,
        },
      },
    },
    silenceForm: {
      strip: {
        labels: [],
      },
    },
    alertAcknowledgement: {
      enabled: false,
      durationSeconds: 900,
      author: "karma / author missing",
      commentPrefix: "",
    },
    staticColorLabels: ["job"],
    annotationsDefaultHidden: false,
    annotationsHidden: [],
    annotationsVisible: [],
  },
});

const MockAPIResponse = () => {
  const response = EmptyAPIResponse();
  response.grids = [
    {
      labelName: "",
      labelValue: "",
      alertGroups: [
        MockAlertGroup(
          { alertname: "foo" },
          [MockAlert([], { instance: "foo" }, "suppressed")],
          [],
          { cluster: "dev" },
          {}
        ),
      ],
      stateCount: {
        unprocessed: 1,
        suppressed: 2,
        active: 3,
      },
    },
  ];
  response.totalAlerts = 25;
  return response;
};

const MockSilenceResponse = (cluster, count) => {
  let silences = [];
  for (var index = 1; index <= count; index++) {
    const silence = MockSilence();
    silence.id = `silence${index}`;
    silences.push({
      cluster: cluster,
      alertCount: 123,
      isExpired: index < 3,
      silence: silence,
    });
  }
  return silences;
};

export { EmptyAPIResponse, MockAPIResponse, MockSilenceResponse };
