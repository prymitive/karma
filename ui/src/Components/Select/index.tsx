import { FC } from "react";

import { components, MenuProps } from "react-select";

import { DropdownSlide } from "Components/Animations/DropdownSlide";
import { OptionT, MultiValueOptionT } from "Common/Select";

export const AnimatedMenu: FC<MenuProps<OptionT, false>> = (props) => {
  return (
    <DropdownSlide in unmountOnExit>
      <components.Menu {...props}>{props.children}</components.Menu>
    </DropdownSlide>
  );
};

export const AnimatedMultiMenu: FC<MenuProps<MultiValueOptionT, true>> = (
  props
) => {
  return (
    <DropdownSlide in unmountOnExit>
      <components.Menu {...props}>{props.children}</components.Menu>
    </DropdownSlide>
  );
};
