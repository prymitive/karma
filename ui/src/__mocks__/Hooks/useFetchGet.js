import { useState, useEffect, useCallback } from "react";

import { MockAPIResponse, MockSilenceResponse } from "__mocks__/Fetch";

const MockFetchStats = {
  getCalls: [],
  get calls() {
    return this.getCalls;
  },
  wasCalled(uri) {
    this.getCalls.push(uri);
  },
  reset() {
    this.getCalls = [];
    this.mockedData = {
      response: undefined,
      error: undefined,
      isLoading: undefined,
      isRetrying: undefined,
    };
  },
  mockedData: {
    response: undefined,
    error: undefined,
    isLoading: undefined,
    isRetrying: undefined,
  },
  setMockedData(data) {
    this.mockedData = data;
  },
};

const Mock = (uri, { autorun = true, deps = [] } = {}) => {
  const [response, setResponse] = useState(null);
  const [error] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const [isRetrying] = useState(false);

  const cancelGet = useCallback(() => {}, []);

  const get = useCallback(() => {
    MockFetchStats.wasCalled(uri);

    const mockResponses = [
      // matcher name select suggestions
      {
        uri: "./labelNames.json",
        response: ["cluster", "job", "instance"],
      },
      // matcher value select suggestions
      {
        uri: "./labelValues.json?name=cluster",
        response: ["dev", "staging", "prod"],
      },
      // matcher value counters
      {
        re: /^\.\/alerts\.json\?q=/,
        response: MockAPIResponse(),
      },
      // silence browser
      {
        re: /^.\/silences\.json\?/,
        response: MockSilenceResponse("am", 14),
      },
      // filter autocomplete
      {
        uri: "./autocomplete.json?term=cluster",
        response: [
          "cluster=staging",
          "cluster=prod",
          "cluster=dev",
          "cluster!=staging",
          "cluster!=prod",
          "cluster!=dev",
        ],
      },
      {
        re: /^.\/autocomplete\.json\?term=/,
        response: ["foo=bar", "foo=~bar"],
      },
    ];

    for (const m of mockResponses) {
      if (m.re && uri.match(m.re)) {
        setResponse(m.response);
        setIsLoading(false);
        break;
      } else if (m.uri === uri) {
        setResponse(m.response);
        setIsLoading(false);
        break;
      }
    }
  }, [uri]);

  useEffect(() => {
    if (autorun) get();
    // eslint doesn't like ...deps
    // eslint-disable-next-line
  }, [uri, get, cancelGet, autorun, ...deps]);

  return {
    response:
      MockFetchStats.mockedData.response !== undefined
        ? MockFetchStats.mockedData.response
        : response,
    error:
      MockFetchStats.mockedData.error !== undefined
        ? MockFetchStats.mockedData.error
        : error,
    isLoading:
      MockFetchStats.mockedData.isLoading !== undefined
        ? MockFetchStats.mockedData.isLoading
        : isLoading,
    isRetrying:
      MockFetchStats.mockedData.isRetrying !== undefined
        ? MockFetchStats.mockedData.isRetrying
        : isRetrying,
    get,
    cancelGet,
  };
};

const useFetchGet = jest.fn(Mock);
useFetchGet.fetch = MockFetchStats;
useFetchGet.mockReset = () => {
  useFetchGet.mockClear();
  useFetchGet.mockImplementation(Mock);
  MockFetchStats.reset();
};

export { useFetchGet };
