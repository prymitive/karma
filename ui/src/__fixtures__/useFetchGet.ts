import { useState, useEffect, useCallback } from "react";

import { MockAPIResponse, MockSilenceResponse } from "__fixtures__/Fetch";
import { FetchGetResultT } from "Hooks/useFetchGet";
import { APIAlertsResponseT, APIManagedSilenceT } from "Models/APITypes";

type responseT = null | string[] | APIAlertsResponseT | APIManagedSilenceT[];

interface mockedDataT {
  response: undefined | responseT;
  error: undefined | null | string;
  isLoading: undefined | boolean;
  isRetrying: undefined | boolean;
  retryCount: undefined | number;
  get: () => void;
  cancelGet: () => void;
}

interface mockFetchStatsT {
  getCalls: string[];
  readonly calls: string[];
  wasCalled: (uri: string) => void;
  reset: () => void;
  mockedData: mockedDataT;
  setMockedData: (data: mockedDataT) => void;
}

const mockGet = jest.fn();
const mockCancelGet = jest.fn();

const MockFetchStats: mockFetchStatsT = {
  getCalls: [],
  get calls() {
    return this.getCalls;
  },
  wasCalled(uri: string) {
    this.getCalls.push(uri);
  },
  reset() {
    this.getCalls = [];
    this.mockedData = {
      response: undefined,
      error: undefined,
      isLoading: undefined,
      isRetrying: undefined,
      retryCount: undefined,
      get: mockGet,
      cancelGet: mockCancelGet,
    };
  },
  mockedData: {
    response: undefined,
    error: undefined,
    isLoading: undefined,
    isRetrying: undefined,
    retryCount: undefined,
    get: mockGet,
    cancelGet: mockCancelGet,
  },
  setMockedData(data: mockedDataT) {
    this.mockedData = data;
  },
};

const useFetchGetMock = (
  uri: string,
  { autorun = true, deps = [] } = {}
): FetchGetResultT<responseT> => {
  const [response, setResponse] = useState(null as responseT);
  const [error] = useState(null);
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [isRetrying] = useState<boolean>(false);

  const cancelGet = useCallback(() => {}, []);

  const get = useCallback(() => {
    MockFetchStats.wasCalled(uri);

    const mockResponses = [
      // matcher name select suggestions
      {
        uri: "./labelNames.json",
        response: ["cluster", "job", "instance"],
      },
      // matcher value select suggestions
      {
        uri: "./labelValues.json?name=cluster",
        response: ["dev", "staging", "prod"],
      },
      // matcher value counters
      {
        re: /^\.\/alerts\.json\?q=/,
        response: MockAPIResponse(),
      },
      // silence browser
      {
        re: /^.\/silences\.json\?/,
        response: MockSilenceResponse("am", 14),
      },
      // filter autocomplete
      {
        uri: "./autocomplete.json?term=cluster",
        response: [
          "cluster=staging",
          "cluster=prod",
          "cluster=dev",
          "cluster!=staging",
          "cluster!=prod",
          "cluster!=dev",
        ],
      },
      {
        re: /^.\/autocomplete\.json\?term=/,
        response: ["foo=bar", "foo=~bar"],
      },
    ];

    for (const m of mockResponses) {
      if (m.re && uri.match(m.re)) {
        setResponse(m.response);
        setIsLoading(false);
        break;
      } else if (m.uri === uri) {
        setResponse(m.response);
        setIsLoading(false);
        break;
      }
    }
  }, [uri]);

  useEffect(() => {
    if (autorun) get();
    // eslint doesn't like ...deps
    // eslint-disable-next-line
  }, [uri, get, cancelGet, autorun, ...deps]);

  return {
    response:
      MockFetchStats.mockedData.response !== undefined
        ? MockFetchStats.mockedData.response
        : response,
    error:
      MockFetchStats.mockedData.error !== undefined
        ? MockFetchStats.mockedData.error
        : error,
    isLoading:
      MockFetchStats.mockedData.isLoading !== undefined
        ? MockFetchStats.mockedData.isLoading
        : isLoading,
    isRetrying:
      MockFetchStats.mockedData.isRetrying !== undefined
        ? MockFetchStats.mockedData.isRetrying
        : isRetrying,
    retryCount:
      MockFetchStats.mockedData.retryCount !== undefined
        ? MockFetchStats.mockedData.retryCount
        : 0,
    get,
    cancelGet,
  };
};

useFetchGetMock.fetch = MockFetchStats;
useFetchGetMock._mockGet = mockGet;
useFetchGetMock._cancelGet = mockCancelGet;

export { useFetchGetMock };
