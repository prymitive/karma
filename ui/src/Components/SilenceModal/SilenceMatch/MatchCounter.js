import React from "react";
import PropTypes from "prop-types";

import { useObserver } from "mobx-react-lite";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExclamationCircle } from "@fortawesome/free-solid-svg-icons/faExclamationCircle";
import { faSpinner } from "@fortawesome/free-solid-svg-icons/faSpinner";

import { FormatBackendURI, FormatAlertsQ } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { SilenceFormMatcher } from "Models/SilenceForm";
import { TooltipWrapper } from "Components/TooltipWrapper";
import { useFetchGet } from "Hooks/useFetchGet";
import { MatcherToFilter, AlertManagersToFilter } from "../Matchers";

const MatchCounter = ({ silenceFormStore, matcher }) => {
  const filters = [MatcherToFilter(matcher)];
  if (silenceFormStore.data.alertmanagers.length) {
    filters.push(AlertManagersToFilter(silenceFormStore.data.alertmanagers));
  }

  const { response, error, isLoading, isRetrying } = useFetchGet(
    FormatBackendURI("alerts.json?") + FormatAlertsQ(filters)
  );

  return useObserver(() =>
    error ? (
      <TooltipWrapper
        title={`Failed to fetch alerts matching this label: ${error}`}
      >
        <FontAwesomeIcon className="text-danger" icon={faExclamationCircle} />
      </TooltipWrapper>
    ) : (
      <TooltipWrapper title="Number of alerts matching this label">
        <span
          className="badge badge-light badge-pill d-block"
          style={{ fontSize: "85%", lineHeight: "1rem" }}
          data-am={silenceFormStore.data.alertmanagers.length}
        >
          {isLoading ? (
            <FontAwesomeIcon
              icon={faSpinner}
              spin
              className={isRetrying ? "text-danger" : ""}
            />
          ) : (
            Math.max(response.totalAlerts, 0)
          )}
        </span>
      </TooltipWrapper>
    )
  );
};
MatchCounter.propTypes = {
  silenceFormStore: PropTypes.instanceOf(SilenceFormStore).isRequired,
  matcher: SilenceFormMatcher.isRequired,
};

export { MatchCounter };
