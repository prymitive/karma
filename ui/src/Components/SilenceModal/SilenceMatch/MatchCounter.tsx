import { FC } from "react";

import { observer } from "mobx-react-lite";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExclamationCircle } from "@fortawesome/free-solid-svg-icons/faExclamationCircle";
import { faSpinner } from "@fortawesome/free-solid-svg-icons/faSpinner";

import { APIAlertsResponseT } from "Models/APITypes";
import { FormatBackendURI, FormatAlertsQ } from "Stores/AlertStore";
import { SilenceFormStore, MatcherWithIDT } from "Stores/SilenceFormStore";
import { TooltipWrapper } from "Components/TooltipWrapper";
import { useFetchGet } from "Hooks/useFetchGet";
import { MatcherToFilter, AlertManagersToFilter } from "../Matchers";

const MatchCounter: FC<{
  silenceFormStore: SilenceFormStore;
  matcher: MatcherWithIDT;
}> = observer(({ silenceFormStore, matcher }) => {
  const filters = [MatcherToFilter(matcher)];
  if (silenceFormStore.data.alertmanagers.length) {
    filters.push(AlertManagersToFilter(silenceFormStore.data.alertmanagers));
  }

  const { response, error, isLoading, isRetrying } =
    useFetchGet<APIAlertsResponseT>(
      FormatBackendURI("alerts.json?") + FormatAlertsQ(filters)
    );

  return error ? (
    <TooltipWrapper
      title={`Failed to fetch alerts matching this label: ${error}`}
    >
      <FontAwesomeIcon className="text-danger" icon={faExclamationCircle} />
    </TooltipWrapper>
  ) : (
    <TooltipWrapper title="Number of alerts matching this label">
      <span
        className="badge bg-light rounded-pill d-block"
        style={{ fontSize: "85%", lineHeight: "1rem" }}
        data-am={silenceFormStore.data.alertmanagers.length}
      >
        {isLoading || response === null ? (
          <FontAwesomeIcon
            icon={faSpinner}
            spin
            className={isRetrying ? "text-danger" : ""}
          />
        ) : (
          Math.max(response.totalAlerts, 0)
        )}
      </span>
    </TooltipWrapper>
  );
});

export { MatchCounter };
