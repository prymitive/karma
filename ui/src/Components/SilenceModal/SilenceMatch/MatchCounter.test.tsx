import { mount } from "enzyme";

import toDiffableHtml from "diffable-html";

import { useFetchGetMock } from "__fixtures__/useFetchGet";
import { EmptyAPIResponse } from "__fixtures__/Fetch";
import {
  SilenceFormStore,
  NewEmptyMatcher,
  MatcherWithIDT,
} from "Stores/SilenceFormStore";
import { useFetchGet } from "Hooks/useFetchGet";
import { StringToOption } from "Common/Select";
import { MatchCounter } from "./MatchCounter";

let matcher: MatcherWithIDT;
let silenceFormStore: SilenceFormStore;

beforeEach(() => {
  silenceFormStore = new SilenceFormStore();
  matcher = NewEmptyMatcher();
  matcher.name = "foo";
  matcher.values = [StringToOption("bar")];
});

afterEach(() => {
  jest.restoreAllMocks();
});

const MountedMatchCounter = () => {
  return mount(
    <MatchCounter silenceFormStore={silenceFormStore} matcher={matcher} />
  );
};

describe("<MatchCounter />", () => {
  it("matches snapshot", () => {
    const tree = MountedMatchCounter();
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("renders spinner icon while fetching", () => {
    useFetchGetMock.fetch.setMockedData({
      response: null,
      error: null,
      isLoading: true,
      isRetrying: false,
      retryCount: 0,
      get: jest.fn(),
      cancelGet: jest.fn(),
    });

    const tree = MountedMatchCounter();
    expect(tree.find("svg.fa-spinner")).toHaveLength(1);
    expect(tree.find("svg.fa-spinner.text-danger")).toHaveLength(0);
  });

  it("renders spinner icon with text-danger while retrying fetching", () => {
    useFetchGetMock.fetch.setMockedData({
      response: null,
      error: null,
      isLoading: true,
      isRetrying: true,
      retryCount: 0,
      get: jest.fn(),
      cancelGet: jest.fn(),
    });

    const tree = MountedMatchCounter();
    expect(tree.find("svg.fa-spinner.text-danger")).toHaveLength(1);
  });

  it("renders error icon on failed fetch", () => {
    useFetchGetMock.fetch.setMockedData({
      response: null,
      error: "failed",
      isLoading: false,
      isRetrying: false,
      retryCount: 0,
      get: jest.fn(),
      cancelGet: jest.fn(),
    });

    const tree = MountedMatchCounter();
    expect(tree.find("svg.fa-exclamation-circle.text-danger")).toHaveLength(1);
  });

  it("totalAlerts is 0 after mount", () => {
    const r = EmptyAPIResponse();
    r.totalAlerts = 0;
    useFetchGetMock.fetch.setMockedData({
      response: r,
      error: null,
      isLoading: false,
      isRetrying: false,
      retryCount: 0,
      get: jest.fn(),
      cancelGet: jest.fn(),
    });

    const tree = MountedMatchCounter();
    expect(tree.text()).toBe("0");
  });

  it("updates totalAlerts after successful fetch", () => {
    const tree = MountedMatchCounter();
    expect(tree.text()).toBe("25");
  });

  it("sends correct query string for a 'foo=bar' matcher", () => {
    MountedMatchCounter();
    expect(
      (useFetchGet as jest.MockedFunction<typeof useFetchGet>).mock.calls[0][0]
    ).toBe("./alerts.json?q=foo%3Dbar");
  });

  it("sends correct query string for a 'foo=~bar' matcher", () => {
    matcher.isRegex = true;
    MountedMatchCounter();
    expect(
      (useFetchGet as jest.MockedFunction<typeof useFetchGet>).mock.calls[0][0]
    ).toBe("./alerts.json?q=foo%3D~%5Ebar%24");
  });

  it("sends correct query string for a 'foo=~(bar|baz)' matcher", () => {
    matcher.values = [StringToOption("bar"), StringToOption("baz")];
    matcher.isRegex = true;
    silenceFormStore.data.setAlertmanagers([]);
    MountedMatchCounter();
    expect(
      (useFetchGet as jest.MockedFunction<typeof useFetchGet>).mock.calls[0][0]
    ).toBe("./alerts.json?q=foo%3D~%5E%28bar%7Cbaz%29%24");
  });

  it("selecting one Alertmanager instance appends it to the filters", () => {
    silenceFormStore.data.setAlertmanagers([
      {
        label: "am1",
        value: ["am1"],
      },
    ]);
    MountedMatchCounter();
    expect(
      (useFetchGet as jest.MockedFunction<typeof useFetchGet>).mock.calls[0][0]
    ).toBe("./alerts.json?q=foo%3Dbar&q=%40alertmanager%3D~%5E%28am1%29%24");
  });

  it("selecting two Alertmanager instances appends it correctly to the filters", () => {
    silenceFormStore.data.setAlertmanagers([
      {
        label: "am1",
        value: ["am1"],
      },
      {
        label: "am2",
        value: ["am2"],
      },
    ]);
    MountedMatchCounter();
    expect(
      (useFetchGet as jest.MockedFunction<typeof useFetchGet>).mock.calls[0][0]
    ).toBe(
      "./alerts.json?q=foo%3Dbar&q=%40alertmanager%3D~%5E%28am1%7Cam2%29%24"
    );
  });
});
