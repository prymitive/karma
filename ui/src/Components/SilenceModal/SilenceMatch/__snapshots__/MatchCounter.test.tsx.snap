// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MatchCounter /> matches snapshot 1`] = `
"
<div class=\\" tooltip-trigger\\">
  <span class=\\"badge bg-light rounded-pill d-block\\"
        style=\\"font-size: 85%; line-height: 1rem;\\"
        data-am=\\"0\\"
  >
    25
  </span>
</div>
"
`;
